clippy_lints/src/and_then_then_some.rs

```diff
@@ -0,0 +1,164 @@
+    /// let _y = x.clone().and_then(|v| v.starts_with('f').then_some(v));
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x = Some("foo".to_string());
+    /// let _y = x.clone().filter(|v| v.starts_with('f'));
+    /// ```
+    #[clippy::version = "1.81.0"]
+    pub AND_THEN_THEN_SOME,
+    nursery,
+    "detects usage of and_then and then_some that can be replaced by filter"
+}
+
+// note: `Option::filter` is older than `bool::then_some`,
+// so no msrv check is required.
+declare_lint_pass!(AndThenThenSome => [AND_THEN_THEN_SOME]);
+
+impl<'tcx> LateLintPass<'tcx> for AndThenThenSome {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        match expr.kind {
+            ExprKind::MethodCall(_, selfarg, [arg], _) | ExprKind::Call(_, [selfarg, arg]) => {
+                // TODO: check if type of reciever is diagnostic item Option?
+                if is_and_then(cx, expr) {
+                    if let Some((closure_args, predicate)) = then_some_closure_arg(cx, arg) {
+                        show_sugg(cx, expr.span, selfarg, closure_args, predicate);
+                    }
+                }
+            },
+            _ => {},
+        }
+    }
+}
+
+// `|v| X.then_some(v)` -> Some((span"|v|", X))
+fn then_some_closure_arg<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'tcx>) -> Option<(Span, &'tcx Expr<'tcx>)> {
+    match expr.kind {
+        ExprKind::Closure(Closure {
+            fn_decl: FnDecl {
+                inputs: [Ty { hir_id: arg_id, .. }],
+                ..
+            },
+            body,
+            ..
+        }) => {
+            if let Node::Expr(expr) = cx.tcx.hir_node(body.hir_id) {
+                (peel_closure_body(cx, expr, *arg_id)).map(|body| (cx.tcx.hir().span(*arg_id), body))
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn peel_closure_body<'tcx>(
+    cx: &LateContext<'tcx>,
+    expr: &'tcx Expr<'tcx>,
+    closure_arg_id: HirId,
+) -> Option<&'tcx Expr<'tcx>> {
+    match expr.kind {
+        // it would be nice if we could lift { x; y.a() } into { x; y }.a()
+        ExprKind::Block(
+            Block {
+                stmts: [],
+                expr: Some(wrapped_expr),
+                ..
+            },
+            _,
+        ) => peel_closure_body(cx, wrapped_expr, closure_arg_id),
+        ExprKind::MethodCall(_, pred, [arg], _) | ExprKind::Call(_, [pred, arg]) => {
+            if is_then_some(cx, expr) && is_local_defined_at(cx, arg, closure_arg_id) {
+                Some(pred)
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+fn is_local_defined_at(cx: &LateContext<'_>, local: &Expr<'_>, arg_hid: HirId) -> bool {
+    match local.kind {
+        ExprKind::Path(QPath::Resolved(
+            _,
+            Path {
+                res: Res::Local(local_hid),
+                ..
+            },
+        )) => {
```

Comment:
looks like `clippy_utils::path_to_local`, but I'm not sure if this can be replaced

Reply:
i'm much more familiar with the hir and mir types than all the clippy utils, and yeah, you can't use function calls in patterns.

Reply:
It could be extracted from the `match` expression then in the `_` case you put that in the other branch. It'd be cleaner imo too

Reply:
Actually, one step further, rather than doing an `if let` as I suggested above, do smth with `Option::map_or_else` then `is_some_and(identity)` since the `true` case is only when `path_to_local` would return `Some`.